crates/ra_hir/src/code_model.rs

```diff
@@ -1333,6 +1333,11 @@ impl Type {
         Some(adt.into())
     }
 
+    pub fn substs(&self) -> Option<Substs> {
+        let (_adt, subst) = self.ty.value.as_adt()?;
+        Some(subst.clone())
+    }
```

Comment:
cc @flodiebold , not sure if want to expose `Subst` from `hir::Type` or not...

My gut feeling is that we want to keep this private, but I don't have a good understanding here...

Reply:
> My gut feeling is that we want to keep this private, but I don't have a good understanding here...

Then what is the right way go get all type parameters?

Reply:
> For the current UI, I could imagine just producing a flat list of all the ADTs mentioned in the type (using walk that would be easy)

@flodiebold are you talking about https://github.com/rust-analyzer/rust-analyzer/blob/0035dafbfa563921e2cfe41f5592fc37bff92294/crates/ra_hir_ty/src/lib.rs#L844-L845 or I got something wrong? Looks like TypeWalk is implemented for `Ty`, `Substs`, etc, not for `Type`. 

Could you please give me some pointers to the existing code that does something similar? I tried to understand how inlay hints (via HirDisplay\HirFormatter?) work, but at the moment it is too hard for me :( Using substs was my bets bet.

Reply:
Yes, that's what I meant. You're right that it isn't implemented for `Type`, which makes sense because like `Subst`, it exposes internals of `hir_ty`, which we don't want to do in the `hir` facade.

What I was imagining is adding a method on `Type` for this; either a very specific one that uses `walk` internally and returns a `Vec<Adt>`, or maybe even a full `walk` function (but one that passes `Type`s instead of `Ty`s).

Reply:
I see. I'll try to get simple flat list for the start.
Thanks.